File: javascript/algorithms/binary_search_iterative.js
/*
- 제목: 이진 탐색 (반복)
- 개념: 정렬된 배열에서 탐색 구간을 절반씩 줄여 목표값의 위치를 찾습니다.
- 시간복잡도(필요한 경우): O(log N)
- 주요 변수 의미: arr: 정렬된 입력 배열, target: 찾을 값, left/right: 현재 탐색 구간 경계, mid: 가운데 인덱스
- 실행 흐름: left~right 범위를 유지하며 mid와 target을 비교해 왼쪽/오른쪽 절반만 남깁니다. 또한 입력 준비 → 핵심 함수 호출 → 출력 검증의 3단계로 코드를 읽으면서, 각 단계에서 어떤 변수가 갱신되고 왜 그런 값이 되는지 추적하면 학습 효과가 크게 높아집니다. 특히 출력을 보기 전에 먼저 예상값을 손으로 계산해 본 뒤 실제 결과와 비교해보면 개념이 훨씬 오래 기억됩니다.
- 오픈 카톡방 : https://open.kakao.com/o/gX0WnTCf
- 책 구매 : https://store.kyobobook.co.kr/person/detail/1112299903
*/

function binarySearchIterative(arr, target) {
  let left = 0;
  let right = arr.length - 1;

  while (left <= right) {
    const mid = Math.floor((left + right) / 2);
    if (arr[mid] === target) return mid;
    if (arr[mid] < target) left = mid + 1;
    else right = mid - 1;
  }

  return -1;
}

const numbers = [1, 3, 5, 7, 9, 11, 13];
console.log(`index: ${binarySearchIterative(numbers, 7)}`);

File: cpp/algorithms/binary_search_iterative.cpp
/*
- 제목: 이진 탐색(반복)
- 목적: 정렬된 배열에서 원하는 값을 빠르게 찾는 방법을 익힙니다.
- 개념: 중앙값(mid)과 target을 비교해 탐색 범위를 절반씩 줄입니다.
- 시간복잡도(필요한 경우): O(log N) (N: 배열 길이)
- 핵심 변수 의미: left/right: 현재 탐색 구간의 양끝 인덱스, mid: 가운데 인덱스, target: 찾을 값
- 오픈 카톡방 : https://open.kakao.com/o/gX0WnTCf
- 책 구매 : https://store.kyobobook.co.kr/person/detail/1112299903
- 결과값 설명(학습 포인트): 출력 인덱스가 곧 target 위치입니다. -1이면 배열에 값이 없다는 의미입니다. 또한 이 출력이 왜 나오는지(입력 데이터와 알고리즘/문법 규칙의 대응 관계)를 역으로 설명해보면 학습 효과가 크게 올라갑니다. 가능하면 출력값이 바뀌도록 입력을 직접 수정해 보고, 어떤 규칙 때문에 결과가 달라졌는지 비교해 보세요. [학습 확장] (1) 먼저 현재 입력 기준으로 왜 이 결과가 나오는지 각 변수의 값 변화(예: 인덱스 이동, 누적값 변화, 자료구조 상태 변화)로 추적해 보세요. (2) 같은 로직에서 입력 크기/값 분포를 바꿨을 때 결과와 실행시간이 어떻게 달라지는지 비교하면 시간복잡도 감각을 함께 얻을 수 있습니다. (3) 출력 결과가 정답 조건을 어떻게 만족하는지(예: 정렬 여부, 최단거리 조건, 중복 제거 결과 일치)를 체크리스트로 검증해 보세요.
*/

#include <iostream>
#include <vector>

int binary_search_iterative(const std::vector<int>& arr, int target) {
    int left = 0, right = static_cast<int>(arr.size()) - 1;
    while (left <= right) {
        int mid = left + (right - left) / 2;
        if (arr[mid] == target) return mid;
        if (arr[mid] < target) left = mid + 1;
        else right = mid - 1;
    }
    return -1;
}

int main() {
    std::vector<int> numbers{1,3,5,7,9,11,13};
    std::cout << "index: " << binary_search_iterative(numbers, 7) << "\n";
}
